sound/soc/sof/intel/hda.c

```diff
@@ -517,6 +521,41 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 		return ret;
 	}
 
+	/* need to power-up core before setting-up capabilities */
+	ret = hda_dsp_core_power_up(sdev, HDA_DSP_CORE_MASK(0));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: could not power-up DSP subsystem\n");
+		return ret;
+	}
+
+	/* initialize SoundWire capabilities */
+	ret = hda_sdw_init(sdev, &link_mask);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: SoundWire get caps error\n");
+		hda_dsp_core_power_down(sdev, HDA_DSP_CORE_MASK(0));
+		return ret;
+	}
+
+	if (!pdata->machine && link_mask != 0) {
+		/* SoundWire enabled, revisit machine driver selection */
+		mach = pdata->desc->alt_machines;
+		while (mach) {
+			if (mach->link_mask == link_mask)
+				break;
+			mach++;
+		}
```

Comment:
``While (mach)`` doesn't work. The latest mach of pdata->desc->alt_machines is not null. 
My suggestion will be
```
		for (mach = pdata->desc->alt_machines; mach->link_mask; mach++) {
			if (mach->link_mask == link_mask) {
				dev_dbg(bus->dev,
					"SoundWire machine driver %s topology %\s\n",
					mach->drv_name,
					mach->sof_tplg_filename);
				pdata->machine = mach;
				mach->mach_params.platform = dev_name(sdev->dev);
				pdata->fw_filename = mach->sof_fw_filename;
				pdata->tplg_filename = mach->sof_tplg_filename;
				break;
			}
		}
		if (!mach->link_mask)
			mach = NULL;
```